src/bevy_compat/keyboard.rs

```diff
@@ -0,0 +1,647 @@
+        self.state.hash(state);
+        self.window.hash(state);
+    }
+}
+
+/// This is a double-buffer for the last keys pressed.
+///
+/// - `.0` is current frame's last keys pressed.
+/// - `.1` is the next frame's last keys pressed.
+#[derive(Default)]
+struct LastPress(HashSet<KeyInput>, HashSet<KeyInput>);
+
+impl LastPress {
+    fn swap(&mut self) {
+        std::mem::swap(&mut self.0, &mut self.1);
+    }
+}
+
+fn send_key_events(
+    mut keys: EventReader<KeyEvent>,
+    kitty_enabled: Option<Res<KittyEnabled>>,
+    window: Query<Entity, With<PrimaryWindow>>,
+    mut modifiers: Local<Modifiers>,
+    mut last_pressed: Local<LastPress>,
+    mut keyboard_input: EventWriter<KeyboardInput>,
+) {
+    let bevy_window = window.single();
+    for key_event in keys.read() {
+        if let Some((bevy_event, mods)) = key_event_to_bevy(key_event, bevy_window) {
+            // dbg!(mods, *modifiers);
+            if mods != **modifiers {
+                let delta = mods.symmetric_difference(**modifiers);
+                for flag in delta {
+                    let state = if mods.contains(flag) {
+                        // This flag has been added.
+                        bevy::input::ButtonState::Pressed
+                    } else {
+                        // modifiers.contains(flag)
+                        // This flag has been removed.
+                        bevy::input::ButtonState::Released
+                    };
+                    keyboard_input.send(modifier_to_bevy(
+                        crossterm_modifier_to_bevy_key(flag),
+                        state,
+                        bevy_window,
+                    ));
+                }
+                **modifiers = mods;
+            }
+            if kitty_enabled.is_none() {
+                // Must send the release events ourselves.
+                let wrapped = KeyInput(bevy_event.clone());
+                if let Some(last_press) = last_pressed.0.take(&wrapped) {
+                    // It's being held down. Pass to the next frame.
+                    last_pressed.1.insert(last_press);
+                } else {
+                    last_pressed.1.insert(wrapped);
+                    keyboard_input.send(bevy_event);
+                }
+            } else {
+                keyboard_input.send(bevy_event);
+            }
+        }
+    }
+    for e in last_pressed.0.drain() {
+        keyboard_input.send(KeyboardInput {
+            state: ButtonState::Released,
+            ..e.0
+        });
+    }
+    last_pressed.swap();
+}
+
+/// This is a dummy window to satisfy the [KeyboardInput] struct.
+fn setup_window(
+    mut commands: Commands,
+    // mut window_created: EventWriter<WindowCreated>
+) {
+    // Insert our window entity so that other parts of our app can use them
+    let _bevy_window = commands.spawn(PrimaryWindow).id();
+
+    // Publish to the app that a terminal window has been created
+    // window_created.send(WindowCreated {
```

Comment:
More commented code- can we remove?

Reply:
We can but I feel like we may be violating an expectation though, but our Window is a dummy anyway so maybe it doesn't matter. `bevy_crossterm` does submit this event. When I tried to do it here, I got a panic. I wasn't sure I wanted to add the `WindowPlugin` to resolve it.

Reply:
I would lean towards not including WindowPlugin or sending this event, but I can't pretend to know everything in bevy that depends on it or whether it would cause problems. Would you be amenable to removing it for now, and we can always add it if we run into a specific bug?

Also small thing, but in the interest of avoiding side effects, I tested the examples passing in an explicitly named `DummyWindow` component instead of `PrimaryWindow`, and it worked the same. Probably better just in case we run into present or future bevy assumptions around the PrimaryWindow.